### PR TITLE
zpk2tf and freqz functions implementation

### PR DESCRIPTION
**Pull request by alykhaled**
I've created an implementation for this issue #2969 
I don't know if I wrote the TypeScript definition in `index.d.ts` the right way.
Please let me know if I need to change anything.

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -2566,6 +2566,28 @@ declare namespace math {
...
+     * Signal functions
+     ************************************************************************/
+    /**
+     * Compute the transfer function of a zero-pole-gain model.
+     * @param z Zeroes of the model
+     * @param p Poles of the model
+     * @param k Gain of the model
+     * @returns The transfer function as array of numerator and denominator
+     */
+    zpk2tf<T extends MathCollection>(z: T, p: T, k?: number): T
```
Comment (josdejong) [inline types/index.d.ts:2579]: The return type of the definition needs to be `{ w: T, h: T }` instead of `T`, right?

File: types/index.d.ts [hunk 2]
```diff
@@ -2566,6 +2566,28 @@ declare namespace math {
...
+
+    /**
+     * Calculates the frequency response of a filter given its numerator and denominator coefficients.
+     * @param b The numerator polynomial of the filter
+     * @param a The denominator polynomial of the filter
+     * @param w The range of frequencies in which the response is to be calculated
+     * @returns The frequency response
+     *
+     */
+    freqz<T extends MathCollection>(b: T, a: T, w?: T): T
```
Comment (josdejong) [inline types/index.d.ts:2589]: Same here: wrong return type

File: types/index.d.ts (hunk 1 above)
Comment (josdejong) [inline types/index.d.ts:2579]: An I think `k` needs to be `number | T`

Comment (josdejong) [reply types/index.d.ts:2579]: O no, nevermind I'm confusing `freqz`

File: types/index.d.ts (hunk 2 above)
Comment (josdejong) [inline types/index.d.ts:2589]: `w ` needs to be `number | T`, right?

File: types/index.d.ts [hunk 3]
```diff
@@ -5920,6 +5944,29 @@ declare namespace math {
       a2: MathCollection
     ): MathJsChain<T>
 
+    /*************************************************************************
+     * Signal functions
+     ************************************************************************/
+    /**
+     * Compute the transfer function of a zero-pole-gain model.
+     */
+    zpk2tf<T extends number | MathArray | MathArray[]>(
```
Comment (josdejong) [inline types/index.d.ts:5953]: The idea of the chained functions is that you chain a number and use the chained number as first argument of a function:

```
const result = math.chain(2).add(3).done() // 5
``` 

So, in the chained type definition, the first argument is replaced by the `this` argument, and besides that, all is the same. So it should look something like:

```ts
zpk2tf<T extends MathCollection>(this: MathJsChain<T>, p: T, k?: T): T
```

File: types/index.d.ts [hunk 4]
```diff
@@ -5920,6 +5944,29 @@ declare namespace math {
...
+      this: MathJsChain<T>,
+      z: T,
+      p: T,
+      k?: T
+    ): MathJsChain<MathArray | MathArray[]>
+
+    /**
+     * Calculates the frequency response of a filter given its numerator and denominator coefficients.
+     */
+    freqz<T extends number | MathArray | MathArray[]>(
```
Comment (josdejong) [inline types/index.d.ts:5963]: Same here, the chained definition needs some changes like explained for `zpk2tf`

**Review by josdejong — CHANGES_REQUESTED** [review]
Thanks @alykhaled !

I made inline comments where changes are needed. Functionally the functions work like a charm, though we need some refinements in the documentation and type definitions etc.

File: test/unit-tests/function/signal/freqz.test.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import approx from '../../../../tools/approx.js'
+import math from '../../../../src/defaultInstance.js'
+
+const freqz = math.freqz
+
+describe('freqz', function () {
```
[the remark this answers is not in the document]
Comment (alykhaled) [reply test/unit-tests/function/signal/freqz.test.js:6]: The only case I could think of now is if the user send negative frequencies points (w) and I implemented a testcase for it now

File: types/index.d.ts (hunk 1 above)
Comment (alykhaled) [reply types/index.d.ts:2579]: right

File: types/index.d.ts (hunk 3 above)
Comment (alykhaled) [reply types/index.d.ts:5953]: Thanks for the explantion.
I tried to copy the same format of the other functions.
Should it be something like that 
`    zpk2tf<T extends MathCollection>(
      this: MathJsChain<T>,
      z: T,
      p: T,
      k?: T
    ): MathJsChain<T>
    `

**Comment by alykhaled** [comment]
I think I fixed all the inline comments can you review it ? @josdejong

File: types/index.d.ts (hunk 1 above)
Comment (josdejong) [reply types/index.d.ts]: Ah, yes, I see 👍

File: types/index.d.ts [hunk 6]
```diff
@@ -5920,6 +5944,28 @@ declare namespace math {
...
+      p: T,
+      k?: number
+    ): MathJsChain<T>
+
+    /**
+     * Calculates the frequency response of a filter given its numerator and denominator coefficients.
+     */
+    freqz<T extends number | MathArray | MathArray[]>(
+      this: MathJsChain<T>,
+      b: T,
```
Comment (josdejong) [inline types/index.d.ts:5964]: So, here, `b` should be removed since that is the chained variable `this`, right?

File: types/index.d.ts [hunk 7]
```diff
@@ -5920,6 +5944,28 @@ declare namespace math {
...
+      a: T,
+      w?: T | number
+    ): MathJsChain<object>
```
Comment (josdejong) [inline types/index.d.ts:5967]: I think the output is not a `MathJsChain<object>` but `MathJsChain<{ w: T; h: T }>`, right?

File: types/index.d.ts [hunk 8]
```diff
@@ -2566,6 +2566,28 @@ declare namespace math {
...
+     * Signal functions
+     ************************************************************************/
+    /**
+     * Compute the transfer function of a zero-pole-gain model.
+     * @param z Zeroes of the model
+     * @param p Poles of the model
+     * @param k Gain of the model
+     * @returns The transfer function as array of numerator and denominator
+     */
+    zpk2tf<T extends MathCollection>(z: T, p: T, k?: number): { w: T; h: T }
```
[the remark this answers is not in the document]
Comment (alykhaled) [reply types/index.d.ts:2579]: Yes you are right I confused it with `freqz`

File: src/function/signal/freqz.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    'Matrix, Matrix, Array': function (b, a, w) {
+      const { h } = _freqz(b.valueOf(), a.valueOf(), w)
+      return {
+        h: matrix(h),
+        w: matrix(w)
+      }
+    },
+    'Matrix, Matrix, number': function (b, a, w) {
+      if (w < 0) {
```
Comment (josdejong) [inline src/function/signal/freqz.js:72]: The logic to create an array `w` from a number input is now duplicated. Can you extract this logic in a function and reuse it?

**Comment by alykhaled** [comment]
I see that the build fails because some problems in the `freqz` docs example 
```
+ {
+   h: [
+     Complex {
+       im: 0,
+       re: 0.5
+     },
+     Complex {
+       im: 0.2861153547458193,
+       re: 0.4768589245763655
+     },
+     Complex {
+       im: 0.75,
+       re: 0.25000000000000006
+     },
+     Complex {
+       im: 0.4625859429811135,
+       re: -0.770976571635189
+     }
+   ],
+   w: [
+     0,
+     0.7853981633974483,
+     1.5707963267948966,
+     2.356194490192345
+   ]
+ }
- '{ h: [0.5+0j, 0.47685+0.2861j, 0.2500+0.75j,  -0.770+0.4625j], w: [0, 0.7853981633974483, 1.5707963267948966, 2.356194490192345 ] }'
```
Should I edit the example to be in this format ?
```
Complex {
  im: 0.4625859429811135,
  re: -0.770976571635189
 }
```

File: src/function/signal/freqz.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        throw new Error('w must be a positive number')
+      }
+      const w2 = []
+      for (let i = 0; i < w; i++) {
+        w2.push(i / w * Math.PI)
+      }
+      return _freqz(b, a, w2)
+    },
+    'Matrix, Matrix': function (b, a) {
+      const { w, h } = _freqz(b.valueOf(), a.valueOf())
```
Comment (josdejong) [inline src/function/signal/freqz.js:58]: I think this function invocation of `_freqz` is missing the third argument `w`.

File: src/function/signal/zpk2tf.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+     * See also:
+     *   freqz
+     * @param {Array} z
+     * @param {Array} p
+     * @param {number} k
+     * @return {Array}
+     *
+     */
+  return typed(name, {
+    'Array,Array,number': function (z, p, k) {
```
Comment (josdejong) [inline src/function/signal/zpk2tf.js:32]: Thanks for adding the Matrix signatures to `freqz`. Can you add those too for `zpk2tf`?

File: test/unit-tests/function/signal/zpk2tf.test.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import approx from '../../../../tools/approx.js'
+import math from '../../../../src/defaultInstance.js'
+
+const zpk2tf = math.zpk2tf
+
+describe('zpk2tf', function () {
+  it('should return the transfer function of a zero-pole-gain model', function () {
+    approx.deepEqual(zpk2tf([math.complex(1, 5)], [math.complex(-2, 0), math.complex(-3, 0)], 1), [[math.complex(1, 0), math.complex(-1, -5)], [math.complex(1, 0), math.complex(5, 0), math.complex(6, 0)]])
```
Comment (josdejong) [inline test/unit-tests/function/signal/zpk2tf.test.js:8]: 😂 how wide a monitor do you use?  These tests have a line length up to 335 characters. Can you reformat this so it is max 100 characters or so? That will make it better readable.

File: src/function/signal/freqz.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+export const createFreqz = /* #__PURE__ */ factory(name, dependencies, ({ typed, add, multiply, Complex, divide, matrix }) => {
+  /**
+     * Calculates the frequency response of a filter given its numerator and denominator coefficients.
+     *
+     * Syntax:
+     *    math.freqz(b, a)
+     *    math.freqz(b, a, w)
+     *
+     * Examples:
+     *   math.freqz([1, 2], [1, 2, 3], 4) // returns { h: [0.5+0j, 0.47685+0.2861j, 0.2500+0.75j,  -0.770+0.4625j], w: [0, 0.7853981633974483, 1.5707963267948966, 2.356194490192345 ] }
```
Comment (josdejong) [inline src/function/signal/freqz.js:23]: Can you change the occurrences of `j` with `i`? To keep all documentation consistent. Also, there is a script that tests whether the examples work as advertised, and it doesn't understand `j` for complex numbers.

File: test/unit-tests/function/signal/freqz.test.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import approx from '../../../../tools/approx.js'
+import assert from 'assert'
+import math from '../../../../src/defaultInstance.js'
+
+const freqz = math.freqz
+
+describe('freqz', function () {
```
Comment (josdejong) [inline test/unit-tests/function/signal/freqz.test.js:7]: Can you have a look at the code coverage report? It shows that parts of the function `freqz` are not tested. Can you add tests where needed?

https://app.codecov.io/gh/josdejong/mathjs/pull/2988

**Comment by josdejong** [comment]
@alykhaled Thanks for the updates. I did another review round and made some more inline comments (we're getting there). Can you resolve those please?

File: types/index.d.ts (hunk 6 above)
Comment (alykhaled) [reply types/index.d.ts:5964]: Yes

File: types/index.d.ts (hunk 7 above)
Comment (alykhaled) [reply types/index.d.ts:5967]: Yes

File: test/unit-tests/function/signal/zpk2tf.test.js (hunk 12 above)
Comment (alykhaled) [reply test/unit-tests/function/signal/zpk2tf.test.js:8]: I actually have a wide monitor I didn't notice it was too long 😂. I reformatted it.

**Comment by alykhaled** [comment]
> I expect the solution is to replace the occurences of `j` with `i`, can you try that first?

Still gives me the same error

File: src/function/signal/zpk2tf.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * Syntax:
+     *      math.zpk2tf(z, p, k)
+     *
+     * Examples:
+     *    math.zpk2tf([1, 2], [-1, -2], 1)    // returns [[1, -3, 2], [1, 3, 2]]
+     *
+     * See also:
+     *   freqz
+     * @param {Array} z
```
Comment (josdejong) [inline src/function/signal/zpk2tf.js:25]: When running the tests locally I saw a warning the documentation of `zpk2tf`. Can you please:

1. put an empty line above `@param {Array} z`
2. add a description to all @param and the @return (like you did for `freqz` too)

**Comment by josdejong** [comment]
> > I expect the solution is to replace the occurences of `j` with `i`, can you try that first?
> 
> Still gives me the same error

Thanks for trying it out. I had a look into it, and it looks like the `doc.test.js` script is not able to correctly compare the actual result and the outcome that you've commented right from the example. It will require some effort to make that work properly, therefore I added the function `freqz` to the list with ignored functions for the time being.

I added one more inline comment. So I think the open issues are reduced to:

1. test coverage (I see you're working on that)
2. docs of `zpk2tf`
3. linting errors (run `npm run lint` and/or `npm run format`)

**Comment by alykhaled** [comment]
The only that was no tested is the case when there are no `w` parameter is sent to `freqz` so the default be 512, but I tried to test the size of the returns, do you have another idea ?
```
  it('should return the frequency response of a zero-pole-gain model when not given w parameter', function () {
    const {h,w} = freqz([math.complex(1, 0), math.complex(-1, -5)], [math.complex(1, 0), math.complex(5, 0), math.complex(6, 0)]) 
    approx.deepEqual(h.length, 512)
    approx.deepEqual(w.length, 512)
  })

  it('should return the frequency response of a zero-pole-gain model given b and a as matrix and not given w parameter', function () {
    const b = math.matrix([math.complex(1, 0), math.complex(-1, -5)])
    const a = math.matrix([math.complex(1, 0), math.complex(5, 0), math.complex(6, 0)])
    const {h,w} = freqz(b, a)
    approx.deepEqual(h._size, [512,])
    approx.deepEqual(w._size, [512,])
  })
```

**Comment by josdejong** [comment]
Code coverage is 100% now as far as I can see :)

**Comment by alykhaled** [comment]
Ok I fixed the format now this is last push I hope 😂

**Review by josdejong — APPROVED** [review]

**Comment by josdejong** [comment]
All green, and the code looks good and well taken care of 🎉 . Thanks again @alykhaled , I'll merge your work now.

**Comment by josdejong** [comment]
Published now in `v11.9.0`

**Pull request merged**